gms/gossiper.cc

```diff
@@ -656,9 +656,6 @@ void gossiper::apply_state_locally_without_listener_notification(std::unordered_
     for (auto& x : map) {
         const inet_address& node = x.first;
         const endpoint_state& remote_state = x.second;
-        if (node == this->get_broadcast_address()) {
-            continue;
-        }
```

Comment:
How does this detect the problem? Will it compare the node uuid and throw?

Reply:
It is checked in storage_service::check_for_endpoint_collision which calls gossiper::is_safe_for_bootstrap.  

We allow to bootstrap a new node in only two cases:

1) The node is a completely new node and no state in gossip at all
2) The node has state in gossip and it is already removed from the
cluster either by nodetool decommission or nodetool removenode

Reply:
But check_for_endpoint_collision() is only called from prepare_do_join(). How is apply_state_locally() related?

Reply:
Note apply_state_locally_without_listener_notification() is related, not apply_state_locally. The flow is gossip send shadow round message, apply_state_locally_without_listener_notification is called to apply states sent back from the response.

Reply:
Yes, but I don't see how it's related to check_for_endpoints_collision().

Reply:
I am not sure what is the "it" here. 

We allow applying states from the shadow round replies with the ip address of this local node to detect node with the ip address already exists in the cluster. check_for_endpoints_collision does the actual check. 

Oh, perhaps I see what confuses you now. check_for_endpoint_collision is only used for bootstrap collision check, not for other checks, e.g, restart a node.

Reply:
Yes. So where does this change make a difference?

Reply:
Without this patch, the gossip state of the node with the same ip address will be ignored. The node will always think there is no existing node of this ip address in the cluster.

Reply:
Ok, I'm not sure I understood 100% but will proceed.